File: ToDoList/app/src/main/java/com/davdo/todolist/src/Note.java
package com.davdo.todolist.src;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.room.Entity;

import java.util.Date;
import java.util.UUID;

@Entity
public class Note implements Parcelable {

    private UUID mId;
    private String mTitle;
    private Date mDate;

    private boolean mDone;

    public Note() {
        this(UUID.randomUUID());
    }

    public Note(UUID id) {
        this("", new Date(), false, id);
    }

    public Note(String title, Date date, boolean done) {
        this(title, date, done, UUID.randomUUID());
    }

    public Note(String title, Date date, boolean done, UUID id) {

        mTitle = title;
        mDate = date;
        mDone = done;
        mId = id;
    }


    protected Note(Parcel in) {
        mTitle = in.readString();
        mDone = in.readByte() != 0;
        mDate = new Date(in.readLong());
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(mTitle);
        dest.writeByte((byte) (mDone ? 1 : 0));
        dest.writeLong(mDate.getTime());
    }

    public UUID getId() {
        return mId;
    }

    public void setId(UUID id) {
        mId = id;
    }

    public String getTitle() {
        return mTitle;
    }

    public void setTitle(String title) {
        mTitle = title;
    }

    public Date getDate() {
        return mDate;
    }

    public void setDate(Date date) {
        mDate = date;
    }

    public boolean isDone() {
        return mDone;
    }

    public void setDone(boolean done) {
        mDone = done;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public static final Creator<Note> CREATOR = new Creator<Note>() {
        @Override
        public Note createFromParcel(Parcel in) {
            return new Note(in);
        }

        @Override
        public Note[] newArray(int size) {
            return new Note[size];
        }
    };

}

File: ToDoListKotlin/app/src/main/java/com/davdo/todolistkotlin/db/Note.kt
package com.davdo.todolistkotlin.db

import android.os.Parcel
import android.os.Parcelable
import androidx.room.Entity
import androidx.room.PrimaryKey
import java.util.*

@Entity
data class Note(@PrimaryKey var uuid: UUID) : Parcelable {

	var title: String
	var date: Date
	var done: Boolean

	init {
		this.title = ""
		this.date = Date()
		this.done = false
	}

	constructor(): this(null,null,null,null)

	constructor(title: String, date: Date, done:Boolean) : this(title, date, done, null)

	constructor(title: String?, date: Date?, done:Boolean?, id: UUID?) : this(UUID.randomUUID()) {

		if(title != null) this.title = title
		if(date != null) this.date = date
		if(done != null) this.done = done
		if(id != null) this.uuid = id
	}

	private constructor(parcel: Parcel) : this(parcel.readString(), Date(parcel.readLong()), (parcel.readInt() == 1), UUID.fromString(parcel.readString()))

	override fun writeToParcel(parcel: Parcel, flags: Int) {
		parcel.writeString(title)
		parcel.writeLong(date.time)
		parcel.writeInt(if (done) 1 else 0)
		parcel.writeString(uuid.toString())
	}

	override fun describeContents(): Int {
		return 0
	}

	companion object CREATOR : Parcelable.Creator<Note> {
		override fun createFromParcel(parcel: Parcel): Note {
			return Note(parcel)
		}

		override fun newArray(size: Int): Array<Note?> {
			return arrayOfNulls(size)
		}
	}
}